pkg/querier/store.go

```diff
@@ -92,9 +113,99 @@ outer:
 			}
 			iterators = append(iterators, iterator)
 		}
+
 		result = append(result, iter.NewNonOverlappingIterator(iterators, labels))
 	}
-	return result, nil
+
+	return iter.NewHeapIterator(result, req.Direction), nil
+}
+
+func loadFirstChunks(ctx context.Context, chks map[model.Fingerprint][][]chunkenc.LazyChunk) error {
+	sp, ctx := opentracing.StartSpanFromContext(ctx, "loadFirstChunks")
+	defer sp.Finish()
+
+	// If chunks span buckets, then we'll have different fetchers for each bucket.
+	chksByFetcher := map[*chunk.Fetcher][]*chunkenc.LazyChunk{}
+	for _, lchks := range chks {
+		for _, lchk := range lchks {
+			chksByFetcher[lchk[0].Fetcher] = append(chksByFetcher[lchk[0].Fetcher], &lchk[0])
+		}
+	}
+
+	sp.LogFields(otlog.Int("fetchers", len(chksByFetcher)))
+
+	errChan := make(chan error)
+	for fetcher, chunks := range chksByFetcher {
+		go func(fetcher *chunk.Fetcher, chunks []*chunkenc.LazyChunk) {
+			keys := make([]string, 0, len(chunks))
+			chks := make([]chunk.Chunk, 0, len(chunks))
+			for _, chk := range chunks {
+				keys = append(keys, chk.Chunk.ExternalKey())
+				chks = append(chks, chk.Chunk)
+			}
+			chks, err := fetcher.FetchChunks(ctx, chks, keys)
+			errChan <- err
+			if err != nil {
+				return
+			}
+
+			for i, chk := range chks {
+				chunks[i].Chunk = chk
+			}
+		}(fetcher, chunks)
+	}
+
+	var lastErr error
+	for i := 0; i < len(chksByFetcher); i++ {
+		if err := <-errChan; err != nil {
+			lastErr = err
+		}
+	}
+
+	return lastErr
+}
+
+func partitionBySeriesChunks(chunks [][]chunk.Chunk, fetchers []*chunk.Fetcher) map[model.Fingerprint][][]chunkenc.LazyChunk {
+	chunksByFp := map[model.Fingerprint][]chunkenc.LazyChunk{}
+	metricByFp := map[model.Fingerprint]model.Metric{}
+	for i, chks := range chunks {
+		for _, c := range chks {
+			fp := c.Metric.Fingerprint()
+			chunksByFp[fp] = append(chunksByFp[fp], chunkenc.LazyChunk{Chunk: c, Fetcher: fetchers[i]})
+			delete(c.Metric, "__name__")
+			metricByFp[fp] = c.Metric
+		}
+	}
+
+	result := make(map[model.Fingerprint][][]chunkenc.LazyChunk, len(chunksByFp))
+
+	for fp, chks := range chunksByFp {
+		result[fp] = partitionOverlappingChunks(chks)
+	}
+
+	return result
+}
+
+func partitionOverlappingChunks(chunks []chunkenc.LazyChunk) [][]chunkenc.LazyChunk {
```

Comment:
nitpick: I would like to have a comment, it was hard to follow.

Reply:
Added some comments.